core/trino-main/src/main/java/io/trino/server/protocol/Query.java

```diff
@@ -791,26 +782,34 @@ private static StageStats toStageStats(StageInfo stageInfo)
                 .setPhysicalInputBytes(stageStats.getPhysicalInputDataSize().toBytes())
                 .setFailedTasks(stageStats.getFailedTasks())
                 .setCoordinatorOnly(stageInfo.isCoordinatorOnly())
-                .setSubStages(subStages.build())
-                .build();
-    }
+                .setNodes(countStageAndAddGlobalUniqueNodes(stageInfo, globalUniqueNodes));
 
-    private static Set<String> globalUniqueNodes(StageInfo stageInfo)
-    {
-        if (stageInfo == null) {
-            return ImmutableSet.of();
+        // Recurse into child stages to create their StageStats
+        List<StageInfo> subStages = stageInfo.getSubStages();
+        if (subStages.isEmpty()) {
+            builder.setSubStages(ImmutableList.of());
         }
-        ImmutableSet.Builder<String> nodes = ImmutableSet.builder();
-        for (TaskInfo task : stageInfo.getTasks()) {
-            // todo add nodeId to TaskInfo
-            URI uri = task.getTaskStatus().getSelf();
-            nodes.add(uri.getHost() + ":" + uri.getPort());
+        else {
+            ImmutableList.Builder<StageStats> subStagesBuilder = ImmutableList.builderWithExpectedSize(subStages.size());
+            for (StageInfo subStage : subStages) {
+                subStagesBuilder.add(toStageStats(subStage, globalUniqueNodes));
+            }
+            builder.setSubStages(subStagesBuilder.build());
         }
 
-        for (StageInfo subStage : stageInfo.getSubStages()) {
-            nodes.addAll(globalUniqueNodes(subStage));
+        return builder.build();
+    }
+
+    private static int countStageAndAddGlobalUniqueNodes(StageInfo stageInfo, Set<String> globalUniqueNodes)
+    {
+        List<TaskInfo> tasks = stageInfo.getTasks();
+        Set<String> stageUniqueNodes = new HashSet<>(tasks.size() + (tasks.size() / 2));
+        for (TaskInfo task : tasks) {
+            String nodeId = task.getTaskStatus().getNodeId();
+            stageUniqueNodes.add(nodeId);
+            globalUniqueNodes.add(nodeId);
```

Comment:
Would it be better if we did `globalUniqueNodes.addAll(stageUniqueNodes)` after this loop ?

Reply:
There’s no particular advantage to bulk inserts on HashSet, the implementation just does another loop with another iterator which is probably slightly slower (extra iterator, iterator is over the Set instead of a List, temporal memory locality of access, etc)